### PR TITLE
Update to Tokio 0.2 and futures 0.3

### LINKED ISSUE #276: Update futures-rs to 0.3
https://github.com/rust-lang-nursery/futures-rs/blob/master/CHANGELOG.md#030---2019-11-5

### PR DESCRIPTION
**Pull request by vorot93**
Fixes #276

### CODE REVIEW

File: bincode-transport/src/lib.rs [hunk 1]
```diff
@@ -12,14 +12,7 @@ use async_bincode::{AsyncBincodeStream, AsyncDestination};
...
 use serde::{Deserialize, Serialize};
-use std::{
-    error::Error,
-    io,
-    marker::PhantomData,
-    net::SocketAddr,
-    pin::Pin,
-    task::{Context, Poll},
-};
+use std::{error::Error, io, marker::PhantomData, net::SocketAddr, pin::Pin, task::*};
```
Comment (tikue) [inline bincode-transport/src/lib.rs]: Below you do `task::{Context, Poll}` with no glob import. Shall we stick with one or the other?

Comment (vorot93) [reply bincode-transport/src/lib.rs]: I pruned the imports and settled on `futures::task::*`.

**Comment by vorot93** [comment]
This is still held up by `tokio` and `tokio-serde`.

File: tarpc/Cargo.toml [hunk 2]
```diff
@@ -27,43 +27,42 @@ travis-ci = { repository = "google/tarpc" }
...
 async-bincode = { optional = true, version = "0.4" }
 futures-legacy = { optional = true, version = "0.1", package = "futures" }
-futures-test-preview = { optional = true, version = "0.3.0-alpha" }
+futures-test = { optional = true, version = "0.3" }
 tokio-io = { optional = true, version = "0.1" }
 tokio-tcp = { optional = true, version = "0.1" }
 
-tokio-net = { optional = true, version = "0.2.0-alpha" }
-tokio-serde-json = { optional = true, version = "0.3" }
+tokio-serde = { optional = true, version = "0.5", features = ["json"] }
```
Comment (tikue) [inline tarpc/Cargo.toml]: Does it need `features = [json]` if the json feature is enabled by other the json-transport feature?

File: tarpc/src/json_transport/mod.rs [hunk 3]
```diff
@@ -178,7 +172,7 @@ where
     type Item = io::Result<Transport<TcpStream, Item, SinkItem>>;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        let next = ready!(self.project().incoming.poll_next(cx)?);
+        let next = ready!(Pin::new(&mut self.project().listener.incoming()).poll_next(cx)?);
```
Comment (tikue) [inline tarpc/src/json_transport/mod.rs]: This constructs a new incoming stream for every call to `poll_next` -- that doesn't seem right. I understand that the types have been made private, so we can't name then in the struct, but is this really the recommended workaround?

File: tarpc/src/rpc/client/channel.rs [hunk 4]
```diff
@@ -716,48 +714,38 @@ mod tests {
...
-        {
-            pin_utils::pin_mut!(resp);
-            let timer = tokio_timer::Timer::default();
-            let handle = timer.handle();
-            let _guard = tokio_timer::set_default(&handle);
-
-            let _ = resp
-                .as_mut()
-                .poll(&mut Context::from_waker(&noop_waker_ref()));
+        let h = tokio::spawn(async move {
```
Comment (tikue) [inline tarpc/src/rpc/client/channel.rs]: What is the spawn doing?

File: tarpc/src/rpc/transport/channel.rs [hunk 5]
```diff
@@ -89,7 +89,8 @@ mod tests {
     use log::trace;
     use std::io;
 
-    #[tokio::test]
+    #[cfg(feature = "tokio1")]
```
Comment (tikue) [inline tarpc/src/rpc/transport/channel.rs]: Tokio is in the dev dependencies right? Maybe this just needs to be `#[cfg(test)]`

File: tarpc/tests/service_functional.rs [hunk 6]
```diff
@@ -109,64 +109,3 @@ async fn concurrent() -> io::Result<()> {
 
     Ok(())
 }
-
-#[tarpc::service(derive_serde = false)]
-trait InMemory {
```
Comment (tikue) [inline tarpc/tests/service_functional.rs]: Does removing these tests mean it's no longer possible to ensure tarpc works in a single threaded context?

**Pull request merged**